### DISCUSSION TITLE
Adding documentation like an Architecture.md to document internals

### DISCUSSION THREAD

**Discussion by hassec** [pydantic/pydantic#10166] [Feature Request]
I think that pydantic has an awesome set of docs to get people started.
Those docs made it really easy to get started with pydantic, and I've been using it since its v1 times.

What I've struggled with, is that once one starts trying to use pydantic for more complex scenarios that require extensions/modifications, one needs to know more about its internals.

For example the discussions about polymorphic classes, in this [thread](https://github.com/pydantic/pydantic/discussions/7008#discussioncomment-7966976) and related other threads touches complex things like `__get_pydantic_core_schema__` and schema references. 

What I'm trying to get to is a better understanding of the processes:

1. Class definition -> .. -> schema construction -> .. -> finalized model
2. Creation of of an instance from e.g. __new__ -> validated instance

These take rather complicated paths that have various points where one can hook in for modification.
A big picture flow chart of what happens could make clearer, where users have customization options.   

I'm curious to understand if there are already existing docs that give an overview over these processes. And/or if the team has considered adding something like an `ARCHITECTURE.md` that would give a bit of a bird's eye overview of those processes. (I've enjoyed reading [this blog post](https://matklad.github.io//2021/02/06/ARCHITECTURE.md.html) about scoping such a document)

As always, thanks again to everyone involved in creating such an awesome library :pray:

**Comment by Viicos** [comment]
Architecture docs got an initial draft here: https://github.com/pydantic/pydantic/pull/10077. This still WIP, but I hope I'll be able to continue working on it. Thanks for the blog post link, looks interesting.

**Reply by hassec** [reply]
@Viicos Thanks for the quick response, I had not seen this PR yet, so I am stoked to see that there is already work going into this kind of documentation.

**Reply by hassec** [reply]
Also found the docs in the `dev` version now, so leaving [a link](https://docs.pydantic.dev/dev/architecture/) for future readers of this discussion. 

Thanks again @Viicos, I'll be sure to check back on that page as additions are made. The current information was already very helpful :)
